suite/tests/api/opnd-a64.c

```diff
@@ -303,21 +307,479 @@ test_opnd_invert_immed_int()
+    }
+
+/* Map SVE element sizes to opnd_size_t */
+#define ELSZ_B OPSZ_1
+#define ELSZ_H OPSZ_2
+#define ELSZ_S OPSZ_4
+#define ELSZ_D OPSZ_8
+
+#define EXPECT(...)                                \
+    app_pc addresses[] = { __VA_ARGS__ };          \
+    vector_address_test_expectation_t expected = { \
+        addresses,                                 \
+        sizeof(addresses) / sizeof(app_pc),        \
+        false,                                     \
+    }
+
+#define VEC_ADDR_TEST(op, pg, mask, create_mem_opnd, decl_expect)                       \
+    {                                                                                   \
+        decl_expect;                                                                    \
+        expected.is_write = op_is_write(OP_##op);                                       \
+        mc.svep[pg].u32[0] = mask;                                                      \
+        opnd_t mem_opnd = create_mem_opnd;                                              \
+        opnd_set_size(&mem_opnd, op_mem_size(OP_##op));                                 \
+        instr_t *instr = INSTR_CREATE_##op##_sve_pred(                                  \
+            drcontext,                                                                  \
+            opnd_create_reg_element_vector(DR_REG_Z31,                                  \
+                                           opnd_get_vector_element_size(mem_opnd)),     \
+            opnd_create_predicate_reg(DR_REG_P0 + pg, false), mem_opnd);                \
+        test_compute_vector_address_helper(drcontext, instr, &mc, &expected, __LINE__); \
+        instr_destroy(drcontext, instr);                                                \
+        mc.svep[pg].u32[0] = 0xffffffff;                                                \
+    }
+
+#define SCALAR_PLUS_VECTOR(xn, zm, el_size, extend, scale)                             \
+    opnd_create_vector_base_disp_aarch64(DR_REG_X0 + xn, DR_REG_Z0 + zm, el_size,      \
+                                         DR_EXTEND_##extend, scale > 0, 0, 0, OPSZ_NA, \
+                                         scale)
+
+    /* Test all the scalar+vector addressing modes.
+     * The opcode used in the instruction shouldn't make a difference to the address
+     * calculation, so these tests cover all addressing modes but not all
+     * (opcode, addressing mode) combinations.
+     */
+
+    /* 32-bit scaled offset [<Xn|SP>, <Zm>.S, <mod> #N] */
+    VEC_ADDR_TEST(ld1h, /*pg=*/0, 0x11111111,
+                  SCALAR_PLUS_VECTOR(SCALAR_BASE_REG, INDEX_REG_S, ELSZ_S, UXTW, 1),
+                  EXPECT((app_pc)0x8000000000020000, (app_pc)0x8000000000040000,
+                         (app_pc)0x8000000000060000, (app_pc)0x8000000000080000,
+                         (app_pc)0x80000001fffe0000, (app_pc)0x80000001fffa0000,
+                         (app_pc)0x80000001fff80000, (app_pc)0x80000001fff60000));
```

Comment:
Where are these numbers coming from? Are they computations on the `mc` above?  But isn't this just a load?  Should these all be named constants to A) share them and B) make it clear where they're coming from?  Maybe it's that this is toward the end of a long review but I am having a hard time figure out whether these are correct as I'm not sure where they came from.

Reply:
These are the address values that `instr_compute_address_ex()` should return for the specified instruction (for the test above `ld1h z31.s, p0/z, [x0, z1.s, uxtw #1]`) and `mc` register values.

This test is hardcoded to use a 256-bit vector length and this instruction uses 32-bit elements so there are 8 addresses.
The base register `x0` has the value `0x8000000000000000`
Taking the first expected address as an example, element 0 of `z1` is `0x00010000` so the offset is
`0x00010000 uxtx #1 == zero_extend_32bit(0x00010000) << 1 == 0x00020000`
so the expected address is `0x8000000000000000 + 0x00020000 = 0x8000000000020000`.